DIPs/DIPxxxx.md

```diff
@@ -0,0 +1,233 @@
+* Templates lose auto inference. As the attributes are applied to *everything* in the scope. This means they are also applied to templates and those templates are thus required to follow the rules of those attributes. Since templates can auto infer these attributes, the desired effect would be to simply have the template continue to be auto inferred. The function calling the template would then enforce any rules that would need to be applied.
+* Attribute bloat at the aggregate scope. Since some attributes don't propagate through all scopes they must be reapplied for each new aggregate scope. This can become tedious and add extra bloat where it shouldn't be necessary.
+* Lack of inverse effects for function attributes leads to odd reordering instead of maintaining the logical order of function definitions affecting readability. The `@attribute:` then appears somewhere in the middle of a large amount of code making it less visible and apparent of what the intention is.
+
+## Prior Work
+
+[DIP1029: Add throw as a function attribute.](https://github.com/dlang/DIPs/blob/master/DIPs/DIP1029.md)
+
+[DIP10XX: Make nothrow the default.](https://github.com/dlang/DIPs/blob/master/DIPs/DIP1029.md)
+
+[DIP1028: Make @safe the default.](https://github.com/dlang/DIPs/blob/master/DIPs/DIP1028.md)
+
+
+## Description
+
+The first part that is required is adding inverses to attributes which do not have one currently. This means introducing new attributes `throw`, `impure`, and `@gc` which correspond to have the inverse effect of the following respectively `nothrow`, `pure`, and `@nogc`. If the default of a module is set to `@nogc`, without the attribute `@gc` that entire module wouldn't be capable of using the GC, even if there may be one or two outlying functions that require it.
+
+```D
+nothrow {
+    throw void foo() {
+        throw new Exception(); // ok is throw
+    }
+}
+
+pure {
+    impure void foo() {
+        __gshared int a; // ok is impure
+    }
+}
+
+@nogc {
+    @gc void foo() {
+        int[] a = new int[10]; // ok is @gc
+    }
+}
+```
+
+Should an attribute and it's inverse appear explicitly declared on the same symbol, it shall be an error.
+
+```D
+throw nothrow void foo(); // error conflicting attributes
+pure impure   void foo(); // "
+@gc @nogc     void foo(); // "
+
+throw:
+    nothrow foo(); // ok, is nothrow
+
+impure:
+    pure void foo(); // ok, is pure
+
+@gc:
+    @nogc void foo(); // ok, is @nogc
+```
+
+The next part is to allow the default to be set at the module level. This would be done by adding the attribute infront of the `module` declaration at the beginning of the source file. The attributes that would be considered are as follows: `@safe`, `@trusted`, `@system`, `nothrow`, `throw`, `pure`, `impure`, `@nogc`, and `@gc`.
```

Comment:
>infront

*in front

> The attributes that would be considered are as follows

Maybe mention why other attributes aren't considered. Some people have suggested `immutable` should be default, and you can currently do `immutable:` with the same problems as `pure:`. Some considerations:
- `private module ` looks like a module that is private, not a module with `private` on by default.
- User defined attributes cannot be inverted since the compiler doesn't know what the opposite of a UDA is.

Reply:
> Some people have suggested immutable should be default, and you can currently do immutable: with the same problems as pure:.

Variables don't suffer from attribute bloat like functions do.

> private module looks like a module that is private, not a module with private on by default.

It isn't something that would be supported. This would be a case where it'd make sense to extend `private:` to function like `@safe/@trusted/@system`. If you have `private struct Blah { void foo(); }`, it doesn't make sense why `Blah` would be private but `Blah.foo` would be public. You can't use it without Blah.

> User defined attributes cannot be inverted since the compiler doesn't know what the opposite of a UDA is.

Since UDAs don't do anything specific, they just exist. You can do something like 
```D
@MyUda:

@ignoreUda!MyUda void foo();

void bar();
```
Then the processing code that uses MyUda could also look to see if it is ignored. There are a number of ways to do this, and it's possible because the functionality of what a UDA does is controlled in the user's source code.